solutions/usaco-790.mdx

```diff
@@ -2,14 +2,118 @@
 id: usaco-790
 source: USACO Gold 2018 January
 title: Cow At Large
-author: Qi Wang
+author: Qi Wang, Jesse Choe
 ---
 
 [Official Analysis](http://www.usaco.org/current/data/sol_atlarge_gold_jan18.html)
 
 **Time Complexity:** $\mathcal{O}(N)$
 
+Suppose that a given farmer reaches Bessie at barn $i$; this is only possible if that farmer reached barn $i$ either: 
+
+* At the same time as Bessie
+* At an earlier time than Bessie
+
+This means that we can treat this problem as a shortest path problem, where the farmer reaches barn $i$ at the same time or earlier than Bessie if $\texttt{dist1[i]} \leq \texttt{dist2[i]}$. Here, $\texttt{dist1[i]}$ represents the shortest distance between any given farmer and the $i$-th barn, and $\texttt{dist2[i]}$ represents the shortest distance between Bessie and the $i$-th barn.
+
+We can run BFS to find $\texttt{dist1[i]}$ and $\texttt{dist2[i]}$. Once we find these shortest distances, we can simulate Bessies' movement by "catching" her when $\texttt{dist1[i]} \leq \texttt{dist2[i]}$. Once Bessie is caught, we can stop her movement and increase the total number of farmers by 1. This greedily finds the minimum number of farmers, since we only store the "optimal" farmers in our answer. 
+
 <LanguageSection>
+<CPPSection>
+```cpp
+#include <bits/stdc++.h>
+
+using namespace std;
+using ll = long long;
+using vl = vector<ll>;
+using pl = pair<ll, ll>;
+
+#define pb push_back
+#define f first
+#define s second
+
+const ll MAXN = 1e5, INF = 1e9;
+
+vl adj[MAXN];
+
+int main(){
+	freopen("atlarge.in", "r", stdin);
+	freopen("atlarge.out", "w", stdout);
+	ll n, k; 
+	cin >> n >> k; 
+	--k;
+	for(int i = 0; i < n - 1; i++){
+		ll a, b; 
+		cin >> a >> b; 
+		--a, --b;
+		adj[a].pb(b), adj[b].pb(a);
+	}
+	/* 
+	 * dist2 represents the shortest path from 
+	 * Bessie's location to barn i
+	 * dist1 represents the shortest path from 
+	 * any given farmers' location to barn i
+	 */
+	vl dist2(n, INF), dist1(n, INF);
+	dist2[k] = 0;
+	queue<ll> q; 
+	q.push(k);
+	// BFS to find the shortest path from node k to node i
+	while (!q.empty()) {
+		ll cur = q.front(); 
+		q.pop();
+		for (ll u: adj[cur]) {
+			if (dist2[cur] + 1 < dist2[u]) {
+				dist2[u] = dist2[cur] + 1;
+				q.push(u);
+			}
+		}
+	}
+	for (int i = 0; i < n; i++) {
+		if (adj[i].size() == 1) {
+			q.push(i);
+			dist1[i] = 0;
+		}
+	}
+	/* 
```

Comment:
use whitespace to separate unrelated blocks of code

Reply:
Isn't that what I did :stuck_out_tongue: ? Wdym

Reply:
don't see any blank lines

Reply:
ok hold on you just added them nvm